src/nvim/linematch.c

```diff
@@ -354,34 +335,67 @@ size_t linematch_nbuffers(const char **diff_blk, const int *diff_len, const size
   size_t memsize_decisions = 0;
   for (size_t i = 0; i < ndiffs; i++) {
     assert(diff_len[i] >= 0);
-    memsize *= i == 0 ? 2 : (size_t)(diff_len[i] + 1);
+    memsize *= (size_t)(diff_len[i] + 1);
     memsize_decisions += (size_t)diff_len[i];
   }
 
   // create the flattened path matrix
   diffcmppath_T *diffcmppath = xmalloc(sizeof(diffcmppath_T) * memsize);
```

Comment:
this is where it will crash, `memsize` gets much larger. any idea on a trigger to disable the grouping optimization when some limit is exceeded? something like "if `memsize` gets larger than some value, disable the grouping optimization and use the space optimization"